hwcomposer/wayland-hwc.cpp

```diff
@@ -772,12 +738,11 @@ keyboard_handle_enter(void *data, struct wl_keyboard *,
     display->keyboard_enter_serial = serial;
 
     std::scoped_lock lock(display->windowsMutex);
-    if (display->windows.find(surface) == display->windows.end())
-        return;
-
-    struct window *window = display->windows[surface];
+    auto window = reinterpret_cast<struct window *>(wl_surface_get_user_data(surface));
```

Comment:
This sounds like a KWin bug, but I'm getting a NULL surface which makes the `wl_surface_get_user_data call` segfault.

Here's how to reproduce:
* Disable multi-windows mode and subsurface mode
* Open 2 app windows
* Open the Waydroid full-ui window
* Crash

Reply:
Digging into the libwayland code, this does not (necessarily) sound like a KWin bug. libwayland validates that the compositor only sends null objects when it is allowed by the protocol, thus KWin sends an object id for an object that was already destroyed.
Either KWin sends this event when it does not yet know about the destruction of this object or it willingly sends this event for a destroyed surface (which would probably be a spec violation).
Either way we have to check for a NULL surface since the first case is valid.